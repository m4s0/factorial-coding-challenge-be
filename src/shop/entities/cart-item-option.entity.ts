import {
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from 'typeorm';
import { CartItem } from './cart-item.entity';
import { ProductOption } from './product-option.entity';

@Entity()
export class CartItemOption {
  @PrimaryGeneratedColumn('uuid')
  declare id: string;

  @ManyToOne(() => CartItem, (cartItem) => cartItem.itemOptions, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  declare cartItem: CartItem;

  @Column()
  declare cartItemId: string;

  @ManyToOne(() => ProductOption, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  declare option: ProductOption;

  @Column()
  declare optionId: string;

  @CreateDateColumn()
  declare createdAt: Date;

  @UpdateDateColumn()
  declare updatedAt: Date;
}
