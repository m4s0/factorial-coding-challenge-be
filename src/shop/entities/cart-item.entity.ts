import {
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from 'typeorm';
import { CartItemOption } from '@Shop/entities/cart-item-option.entity';
import { Cart } from './cart.entity';
import { Product } from './product.entity';

@Entity()
export class CartItem {
  @PrimaryGeneratedColumn('uuid')
  declare id: string;

  @ManyToOne(() => Cart, (cart) => cart.items, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  declare cart: Cart;

  @Column()
  declare cartId: string;

  @ManyToOne(() => Product, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  declare product: Product;

  @Column()
  declare productId: string;

  @Column({ type: 'int', default: 1 })
  declare quantity: number;

  @OneToMany(
    () => CartItemOption,
    (cartItemOption) => cartItemOption.cartItem,
    {
      cascade: true,
    },
  )
  declare itemOptions: CartItemOption[];

  @CreateDateColumn()
  declare createdAt: Date;

  @UpdateDateColumn()
  declare updatedAt: Date;

  @Column({ type: 'decimal', precision: 10, scale: 2 })
  declare price: number;

  @Column({ type: 'decimal', precision: 10, scale: 2 })
  declare totalPrice: number;
}
