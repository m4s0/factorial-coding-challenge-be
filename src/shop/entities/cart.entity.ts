import {
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  OneToMany,
  OneToOne,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from 'typeorm';
import { User } from '@User/entities/user.entity';
import { CartItem } from './cart-item.entity';

@Entity()
export class Cart {
  @PrimaryGeneratedColumn('uuid')
  declare id: string;

  @OneToOne(() => User, (user) => user.cart, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  declare user: User;

  @Column()
  declare userId: string;

  @OneToMany(() => CartItem, (item) => item.cart, {
    cascade: true,
  })
  declare items: CartItem[];

  @CreateDateColumn()
  declare createdAt: Date;

  @UpdateDateColumn()
  declare updatedAt: Date;

  declare totalPrice?: number;
}
