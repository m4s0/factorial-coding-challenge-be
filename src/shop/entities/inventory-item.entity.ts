import {
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  OneToOne,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from 'typeorm';
import { ProductOption } from './product-option.entity';

@Entity()
export class InventoryItem {
  @PrimaryGeneratedColumn('uuid')
  declare id: string;

  @Column({ type: 'int' })
  declare quantity: number;

  @Column({ default: false })
  declare outOfStock: boolean;

  @OneToOne(
    () => ProductOption,
    (productOption) => productOption.inventoryItem,
    {
      onDelete: 'CASCADE',
    },
  )
  @JoinColumn()
  declare productOption: ProductOption;

  @Column()
  declare productOptionId: string;

  @CreateDateColumn()
  declare createdAt: Date;

  @UpdateDateColumn()
  declare updatedAt: Date;
}
