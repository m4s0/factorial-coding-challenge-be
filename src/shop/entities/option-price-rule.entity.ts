import {
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from 'typeorm';
import { ProductOption } from './product-option.entity';

@Entity()
export class OptionPriceRule {
  @PrimaryGeneratedColumn('uuid')
  declare id: string;

  @Column({ type: 'decimal', precision: 10, scale: 2 })
  declare price: number;

  @ManyToOne(() => ProductOption, (option) => option.priceRules, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  declare targetOption: ProductOption;

  @Column()
  declare targetOptionId: string;

  @ManyToOne(() => ProductOption, (option) => option.priceRules, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  declare dependentOption: ProductOption;

  @Column()
  declare dependentOptionId: string;

  @Column({ default: true })
  declare isActive: boolean;

  @CreateDateColumn()
  declare createdAt: Date;

  @UpdateDateColumn()
  declare updatedAt: Date;
}
