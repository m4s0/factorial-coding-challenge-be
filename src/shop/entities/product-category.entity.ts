import {
  Column,
  CreateDateColumn,
  Entity,
  OneToMany,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from 'typeorm';
import { Product } from './product.entity';

@Entity()
export class ProductCategory {
  @PrimaryGeneratedColumn('uuid')
  declare id: string;

  @Column()
  declare name: string;

  @Column()
  declare description: string;

  @Column({ default: true })
  declare isActive: boolean;

  @OneToMany(() => Product, (product) => product.category)
  declare products: Product[];

  @CreateDateColumn()
  declare createdAt: Date;

  @UpdateDateColumn()
  declare updatedAt: Date;
}
