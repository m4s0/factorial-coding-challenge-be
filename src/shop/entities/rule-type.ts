export enum RuleType {
  REQUIRES = 'requires',
  EXCLUDES = 'excludes',
  ONLY_ALLOWS = 'only_allows',
}
