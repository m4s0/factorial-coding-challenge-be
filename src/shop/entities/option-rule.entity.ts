import {
  Column,
  CreateDateColumn,
  Entity,
  JoinColumn,
  ManyToOne,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
} from 'typeorm';
import { RuleType } from './rule-type';
import { ProductOption } from './product-option.entity';

@Entity()
export class OptionRule {
  @PrimaryGeneratedColumn('uuid')
  declare id: string;

  @Column({
    type: 'enum',
    enum: RuleType,
  })
  declare ruleType: RuleType;

  @ManyToOne(() => ProductOption, (option) => option.rulesAsCondition, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  declare ifOption: ProductOption;

  @Column()
  declare ifOptionId: string;

  @ManyToOne(() => ProductOption, (option) => option.rulesAsResult, {
    onDelete: 'CASCADE',
  })
  @JoinColumn()
  declare thenOption: ProductOption;

  @Column()
  declare thenOptionId: string;

  @Column({ default: true })
  declare isActive: boolean;

  @CreateDateColumn()
  declare createdAt: Date;

  @UpdateDateColumn()
  declare updatedAt: Date;
}
