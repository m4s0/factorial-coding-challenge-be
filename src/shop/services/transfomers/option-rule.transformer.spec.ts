import { OptionRule } from '@Shop/entities/option-rule.entity';
import { RuleType } from '@Shop/entities/rule-type';
import { transformOptionRule } from './option-rule.transformer';

describe('transformOptionRule', () => {
  it('should transform option rule correctly', () => {
    const optionRule = {
      id: 'rule123',
      ruleType: RuleType.REQUIRES,
      ifOptionId: 'option1',
      ifOption: {
        id: 'option1',
        name: 'Size',
        displayName: 'Product Size',
        isActive: true,
        createdAt: new Date('2024-01-01'),
        updatedAt: new Date('2024-01-02'),
      },
      thenOptionId: 'option2',
      thenOption: {
        id: 'option2',
        name: 'Color',
        displayName: 'Product Color',
        isActive: true,
        createdAt: new Date('2024-01-01'),
        updatedAt: new Date('2024-01-02'),
      },
      isActive: true,
      createdAt: new Date('2024-01-01'),
      updatedAt: new Date('2024-01-02'),
    } as OptionRule;

    const result = transformOptionRule(optionRule);

    expect(result).toMatchSnapshot();
  });
});
