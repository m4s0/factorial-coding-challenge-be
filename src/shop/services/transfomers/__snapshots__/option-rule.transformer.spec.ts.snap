// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformOptionRule should transform option rule correctly 1`] = `
{
  "createdAt": "2024-01-01T00:00:00.000Z",
  "id": "755cf8ae-950c-4be2-9a20-1fb6576b745e",
  "ifOption": {
    "basePrice": NaN,
    "createdAt": "2024-01-01T00:00:00.000Z",
    "displayName": "Product Size",
    "id": "4636127e-769e-4136-9a6a-8fddb7ca9692",
    "isActive": true,
    "name": "Size",
    "updatedAt": "2024-01-02T00:00:00.000Z",
  },
  "ifOptionId": "4636127e-769e-4136-9a6a-8fddb7ca9692",
  "isActive": true,
  "ruleType": "requires",
  "thenOption": {
    "basePrice": NaN,
    "createdAt": "2024-01-01T00:00:00.000Z",
    "displayName": "Product Color",
    "id": "fb5d0e88-8014-4883-82d5-94d0af8e723f",
    "isActive": true,
    "name": "Color",
    "updatedAt": "2024-01-02T00:00:00.000Z",
  },
  "thenOptionId": "fb5d0e88-8014-4883-82d5-94d0af8e723f",
  "updatedAt": "2024-01-02T00:00:00.000Z",
}
`;
